### ISSUE TITLE
How does TFX support Online learning?

### ISSUE THREAD

**Issue by liutongxuan** [tensorflow/tfx#2520]
Online training need to run Trainer(and pre-components) every minutes,  and incrementally push to serving. Any idea of TFX support such scenario.

**Comment by rmothukuru** [comment]
@liutongxuan,
Can you please look into this documentation and let us know if it helps. Thanks!

**Comment by liutongxuan** [comment]
@rmothukuru @zhitaoli  it seems like continuous learning (https://www.usenix.org/system/files/opml19papers-baylor.pdf), Our requirement is that samples are from realtime online service and save model need to be generated incrementally based on realtime data every minute.

**Comment by liutongxuan** [comment]
But we found TFX need to save intermediate results into meta store for each component, seems not good performance in our scenario.

**Comment by steven-mi** [comment]
You could schedule your pipeline with a orchestrating tool e.g. in airflow you could specify to run the DAG every minute.

**Comment by zhitaoli** [comment]
> You could schedule your pipeline with a orchestrating tool e.g. in airflow you could specify to run the DAG every minute.

I don't think we have seen anymore deploying the system in that style yet, including ourselves. While I don't know anything which could break for sure, a couple of things to watch out:
1. watch growth on both artifact storage (GCS/S3/HDFS) and MLMD. Especially the latter, as the volume of executions and artifacts could even hurt system performance.
1. it's one thing of how faster we schedule the pipeline to run, but another thing how fast it can run itself. Even if airflow could schedule a DAG every minute reliably (which needs its own verification), the cost of scheduling and overhead of node execution would kick in at this level. It's good to understand how that affects your goal.
1. many nodes in a typical TFX/ML pipeline could take long to finish. I think our "async" pipeline design in [this RFC](https://github.com/tensorflow/community/pull/253) might be more friendly. We are still figuring out how to bring this to OSS. Happy to engage more.

**Comment by steven-mi** [comment]
I think I maybe misunderstood your goal. I was under the assumption that you wanted to run your pipeline multiple times, which is why I suggests to use the integrated orchestrating tool. 

As far as I understood your post, I think your main problem is that the execution time is not consistence, because the amount of data is not. I don't believe that async mode is a good idea, since it will give you more headache than needed. If one would introduce asynchronously to the pipeline, then you would have the functionality twice, which also would give you additional overhead. Wouldn't that be correct? 

Keeping that in mind, wouldn't be a reduction in datapoint more optional? You could limit your data to a certain amount to ensure consistency throughout every run.

**Comment by tanzhenyu** [comment]
> But we found TFX need to save intermediate results into meta store for each component, seems not good performance in our scenario.

Related to this: https://github.com/kubeflow/pipelines/issues/1053#issuecomment-477443017. Though TFX should support data triggering through ExampleGen.

**Comment by gowthamkpr** [comment]
@liutongxuan You can trigger your pipeline on demand, on a schedule, on the availability of new data, on model performance degradation, on significant changes in the statistical properties of the data, or based on other conditions. Please take a look at how to automate your ML pipeline using Kubeflow, tfx and on GCP [here](You can trigger your pipeline on demand, on a schedule, on the availability of new data, on model performance degradation, on significant changes in the statistical properties of the data, or based on other conditions). Thanks!

**Comment by gowthamkpr** [comment]
Closing this issue as there is no response for a week. Please reply to the above comment and we can open this issue again. Thanks!!!
